lib/portage/dbapi/vartree.py

```diff
@@ -5800,28 +5802,33 @@ def mergeme(
                 # whether config protection or not, we merge the new file the
                 # same way.  Unless moveme=0 (blocking directory)
                 if moveme:
-                    # Create hardlinks only for source files that already exist
-                    # as hardlinks (having identical st_dev and st_ino).
-                    hardlink_key = (mystat.st_dev, mystat.st_ino)
+                    # only replace the existing file if it differs, see #722270
+                    already_merged = os.path.exists(mydest)
```

Comment:
Instead of  ` os.path.exists(mydest)` which adds a redundant stat call, we might use something like this:
```
already_merged = mydstat is not None and stat.S_ISREG(mydstat.st_mode)
```

Reply:
The testcase `portage.tests.emerge.test_config_protect.ConfigProtectTestCase.testConfigProtect` fails when relying on `mydstat`. I'm not sure if the test needs some adaptions or if I didn't handle some special case
```
Traceback (most recent call last):
  File "/usr/lib/python3.11/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/lib/python3.11/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/michi/Workspace/portage/lib/portage/util/_async/ForkProcess.py", line 162, in _bootstrap
    sys.exit(self._run())
             ^^^^^^^^^^^
  File "/home/michi/Workspace/portage/lib/portage/dbapi/_MergeProcess.py", line 237, in _run
    rval = mylink.merge(
           ^^^^^^^^^^^^^
  File "/home/michi/Workspace/portage/lib/portage/dbapi/vartree.py", line 1906, in wrapper
    return f(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/michi/Workspace/portage/lib/portage/dbapi/vartree.py", line 6037, in merge
    retval = self.treewalk(
             ^^^^^^^^^^^^^^
  File "/home/michi/Workspace/portage/lib/portage/dbapi/vartree.py", line 5022, in treewalk
    rval = self._merge_contents(srcroot, destroot, cfgfiledict)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/michi/Workspace/portage/lib/portage/dbapi/vartree.py", line 5322, in _merge_contents
    if self.mergeme(
       ^^^^^^^^^^^^^
  File "/home/michi/Workspace/portage/lib/portage/dbapi/vartree.py", line 5809, in mergeme
    if already_merged and filecmp.cmp(mysrc, mydest, shallow=False):
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/filecmp.py", line 54, in cmp
    s2 = _sig(os.stat(f2))
              ^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/portage.runtests.vv22c1ju/tmpzol2s1nj/etc/A/._cfg0000_regular_a'

```

Reply:
@zmedico Could you take a look when you get a chance?

Reply:
It works like this:
```diff
diff --git a/lib/portage/dbapi/vartree.py b/lib/portage/dbapi/vartree.py
index 317cf327a..13b201a3b 100644
--- a/lib/portage/dbapi/vartree.py
+++ b/lib/portage/dbapi/vartree.py
@@ -5544,6 +5544,8 @@ class dblink:
                     destmd5,
                     mydest_link,
                 )
+                if protected and moveme:
+                    mydmode = None
 
             zing = "!!!"
             if not moveme:
@@ -5584,6 +5586,7 @@ class dblink:
                         msg.append("")
                         self._eerror("preinst", msg)
                         mydest = newdest
+                        mydmode = None
 
                 # if secondhand is None it means we're operating in "force" mode and should not create a second hand.
                 if (secondhand is not None) and (not os.path.exists(myrealto)):
@@ -5797,6 +5800,7 @@ class dblink:
                     msg.append("")
                     self._eerror("preinst", msg)
                     mydest = newdest
+                    mydmode = None
 
                 # whether config protection or not, we merge the new file the
                 # same way.  Unless moveme=0 (blocking directory)
@@ -6260,7 +6264,7 @@ class dblink:
         Takes file mode and extended attributes into account.
         Should only be used for regular files.
         """
-        if not os.path.exists(mydest):
+        if not (mydmode is not None and stat.S_ISREG(mydmode)):
             return True
 
         if mymode != mydmode:

```